### DISCUSSION TITLE
Generating proto files error

### DISCUSSION THREAD

**Discussion by johannes-do** [foxglove/community#641] [Q&A]
I am trying to fill the proto schemas for foxglove studio and send this data via websockets for visualization. 

As I first test I just copied the foxglove .poroto files and try to build them with cmake. But what troubles me is the import from foxglove e.g. `import "foxglove/Quaternion.proto";`

What happens is that in Pose.pb.cc this define is used: `PROTOBUF_INTERNAL_EXPORT_foxglove_2fQuaternion_2eproto` while the define in Quatrino.pb.h is named like this: `PROTOBUF_INTERNAL_EXPORT_Quaternion_2eproto`

This is my CMakeLists.txt:

```cmake
cmake_minimum_required(VERSION 3.10)
set(LIB_NAME "libfoxglove_interface")

if (NOT Protobuf_FOUND)
  find_package(Protobuf REQUIRED)
  # exposed CMake variables and functions can be found here:
  # https://cmake.org/cmake/help/latest/module/FindProtobuf.html
endif()

# Generate the .h and .cxx files
set(Protobuf_IMPORT_DIRS ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/..)
# file(GLOB_RECURSE PROTO_FILES *.proto)
protobuf_generate_cpp(PROTO_SRCS PROTO_HDRS Vector3.proto Quaternion.proto Pose.proto)

add_library(${LIB_NAME} STATIC
  ${PROTO_SRCS}
  ${PROTO_HDRS}
)
target_include_directories(${LIB_NAME} SYSTEM
  PUBLIC
    $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
    $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/..>
    $<BUILD_INTERFACE:${PROTOBUF_INCLUDE_DIRS}>
)

install(
  TARGETS ${LIB_NAME}
  DESTINATION lib
  COMPONENT TARGET_COMPONENT
)
```

It works fine if I change all the includes to include from root, so for example instead of `import "foxglove/Quaternion.proto";`, `import "Quaternion.proto";`

Is this an issue with my cmake setup? Bug in protobuf? Bug in foxglove proto description?

**Comment by jtbandes** [comment]
I'm not very familiar with the protobuf cmake rules, but it's common for proto files to use imports like this. When using protoc at the command line, `--proto_path` (`-I`) is used to specify the include dir (parent of `foxglove/`) which will allow the imports to work.

**Comment by johannes-do** [comment]
This can be done in cmake with `Protobuf_IMPORT_DIRS` (see docs: https://cmake.org/cmake/help/latest/module/FindProtobuf.html). The imports work, and all the cpp and h files are generated. But then it seems some defines have inconsistent names.

Most likely its not a foxglove issue, but either a cmake, protobuf or user issue. But not 100% sure. And thought maybe someone in the community has already built the foxglove proto files with cmake and came across the same problem.

**Comment by muhammadilahee-a3** [comment]
Take a look at this example how protobuf sources are generated. This might be helpful https://github.com/foxglove/ws-protocol/blob/main/cpp/examples/CMakeLists.txt
